### ISSUE TITLE
Potential Bug in the noise positional embedding

### ISSUE THREAD

**Issue by ahmdtaha** [bwconrad/can#1]
Shouldn't this [line](https://github.com/bwconrad/can/blob/eb6dbe98ab71cceef924b9d05312d4f11edc0440/src/network/decoder.py#L101) be above this [line](https://github.com/bwconrad/can/blob/eb6dbe98ab71cceef924b9d05312d4f11edc0440/src/network/decoder.py#L98)?

Concretely, it should be
```
# Add noise level embedding
if p is not None:
    temp = temp + p[:, None, :]
# Prepend cls token
 x = torch.cat([x[:, :1, :], temp], dim=1)
```

**Comment by bwconrad** [comment]
You're right. Thanks for pointing that out! I've gone ahead and fixed the bug.

**Comment by ahmdtaha** [comment]
Thanks Ben for your implementation. Besides this bug, I have other concerns about your implementation. While your implementation follows the paper strictly, I feel something needs to be fixed. This [line](https://github.com/bwconrad/can/blob/fa8ae54fd3db4158ba114556495548ff8fff1ebd/src/network/decoder.py#L106) embeds both the un-masked and mask tokens! This is indeed how the paper describes CAN method, but it makes little sense. For instance, it breaks from MAE which embeds only un-tmasked tokens; check [this line](https://github.com/facebookresearch/mae/blob/efb2a8062c206524e35e47d04501ed4f544c0ae8/models_mae.py#L174). 

Think about it; If we set the `weight_contrast=0` and `weight_denoise=0`, CAN should become MAE. Yet, the current implementation doesn't! The current implementation's decoder behaves differently from MAE's decoder.

Any luck replicating the paper results?

**Comment by bwconrad** [comment]
Yes, I tried to follow what is written in the paper as close as possible but, as you brought up, I'm unsure about the order of  operations in the decoder since since they do say that the mask tokens are appended to the input before going through the decoder. 

I went ahead and added an argument `--model.decoder_embed_unmasked_token` which when set to `False` will follow MAE and only embed the masked tokens at the very beginning of the decoder. It defaults to `True` which keeps how it was functioning originally. 

I do not have the resources to try and replicate their results so I can't comment on if these changes have any real impact.

**Comment by ahmdtaha** [comment]
Thanks Ben for your input. I will reach out for the authors and see if they can comment.
Thanks again for sharing your implementation.

**Comment by joshr17** [comment]
Hi Ben and Ahmed,

Thanks for the implementation of CAN, and the careful examination of the workflow! We are really happy to see this repo. 

On the issue's original subject matter - CLS then PE, or PE then CLS - Ahmed is correct. The order of operations is:

> `emb <- emb + pe `


then


> `emb <- concat(cls, emb)`

We didn't try with the other order, though I expect it would work fine. Note we use **average pooling readout, not CLS readout** when doing linear probe and fineunting as we found it to work slightly better. 

The second point Ahmed raises is more important. Indeed, when `weight_contrast=0 `and `weight_denoise=0` CAN should become MAE. In particular, we **only embed un-masked tokens** as is done in MAE. This is stated in Section 3.1: "Following He et al. (2022), only the T' unmasked patches are passed to the ViT encoder". Apologies for any confusion over this point, since it is important.

Do let me know if anything is unclear - I am very happy to continue this discussion further as needed.

Best,

Josh

**Comment by bwconrad** [comment]
@joshr17 Hi Josh. Thanks for clarifying our questions. 

About the second point on embedding the masked tokens, we were actually wondering about the ViT decoder, not the encoder as what the part from section 3.1 is referring to. 
 
In MAE, the decoder first embeds its input to a lower dimension and then goes through transformer blocks. The main point of confusion is that in MAE, only the output tokens of the encoder are embedded and then mask tokens are appended to it afterwards. But from what I understand, in CAN the mask tokens are first appended to encoder's output and then it is passed through the decoder where it would go through the embedding layer. 

So the actual question here is when is the decoder's embedding layer applied (assuming this layer is still being used like in MAE).
